### PR TITLE
Assignment2 vibrato

### PR DESCRIPTION
**Pull request by Jiarui-Xu-Gatech**

### CODE REVIEW

**Review by alisonbma — COMMENTED** [review]
- Can consider to implement a test to ensure that m_fDelayInSamples is set to 0 after incrementing in Vibrato->process()
- I don't know if you take this into account, but the wavetable size might end at a sample that is not a full period of the LFO shape. So, it is possible that you index at samples that go over the wavetable and that your vibrato modulation is not periodic.
- Otherwise, setup of the code seems reasonable
- Could be useful to avoid committing all the CMake files to git, makes it hard to review

File: src/MUSI6106Exec/MUSI6106Exec.cpp [hunk 1]
```diff
@@ -16,29 +15,152 @@ void    showClInfo ();
...
+    //============================================================================
+
+    if(argc > 1 && argc != 6)
+    {
+        std::cout << "Usage: " << std::endl;
+        std::cout << "<input audio Path>" << std::endl;
+        std::cout << "<output audio path>" << std::endl;
+        std::cout << "<delay in seconds>" << std::endl;
+        std::cout << "<modulation width in seconds>" << std::endl;
+        std::cout << "modulation frequency in Hz"<< std::endl;
```
Comment (med1844) [inline src/MUSI6106Exec/MUSI6106Exec.cpp:52]: You forgot to add brackets for the last argument.

File: src/MUSI6106Exec/MUSI6106Exec.cpp [hunk 2]
```diff
@@ -16,29 +15,152 @@ void    showClInfo ();
...
+
+
+    while (!phInputAudioFile->isEof())
     {
-        pCRingBuff->putPostInc(1.F*i);
+        // set block length variable
+        long long int iNumFrames = kBlockSize;
+
+        // read data (iNumOfFrames might be updated!)
+        phInputAudioFile->readData(ppfInputAudioData, iNumFrames);
```
Comment (med1844) [inline src/MUSI6106Exec/MUSI6106Exec.cpp:135]: Since `iNumFrames == kBlockSize` always holds true, why not just use `kBlockSize`? Is this some specific design choice?

File: src/Tests/Tests/Test_Vibrato.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+namespace vibrato_test {
+    void CHECK_ARRAY_CLOSE(float* buffer1, float* buffer2, int iLength, float fTolerance)
+    {
+        for (int i = 0; i < iLength; i++)
+        {
+            EXPECT_NEAR(buffer1[i], buffer2[i], fTolerance);
+        }
+    }
+
+    void RingBufferTest(int iRingBuffLength) {
```
Comment (med1844) [inline src/Tests/Tests/Test_Vibrato.cpp:21]: I don't know why these functions are not written in gtest framework like `Test_AudioIO.cpp`. Shouldn't it use `TEST_F` to define test functions? The setup should be done by inheriting `testing::Test` and override `SetUp()` method. The function here does not take any advantage of gtest framework, which does not fulfill the 0st requirement of the assignment.

> 0. [10] Implement the Ringbuffer as specified in exercise 3: ringbuffer with fractional delay but **move your tests to the gtest framework** (see below).

Comment (med1844) [reply src/Tests/Tests/Test_Vibrato.cpp:21]: And I don't know why ring buffer test was put in the namespace of `vibrato_test`?... I also looked around the code base and I don't see anywhere these functions are called and where parameters are fed. As long as these functions are not defined as `TEST_F`, I doubt if they will be found by [const UnitTest::TestList& allTests( UnitTest::Test::GetTestList() );](https://github.com/Jiarui-Xu-Gatech/2022-MUSI6106/blob/assignment2_vibrato/src/Tests/TestExec/Test.cpp#L90). What's more, **each test deletes INPUT FLOAT ARRAYS**. That is to say, after calling each function, the outside caller have to **reallocate** and **re-initialize** these arrays again, which does not make sense at all. It's also deleting something that is not allocated in the function body.

The design here is absolutely terrible and requires full rewrite.

File: src/inc/Lfo.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    float m_fFreqInSamples;
+    float m_fAmplitude;
+    float m_fSampleRate;
+    float m_fCurrIdx;
+
+    Error_t createWaveTable()
+    {
+        for (int i=0; i< m_iWaveTableLength; i++)
+        {
+            m_pfWaveTable ->putPostInc(std::sin(2 * M_PI * i / m_iWaveTableLength));
```
Comment (med1844) [inline src/inc/Lfo.h:76]: Consider use functionality provided by `Synthesis.h` instead.

File: src/inc/Lfo.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        createWaveTable();
+        m_bIsInitialized = true;
+    }
+
+    float process()
+    {
+        float fCurrValue = m_pfWaveTable -> get(m_fCurrIdx);
+        float fIdxInc = m_fFreqInSamples * m_iWaveTableLength / m_fSampleRate ;
+
+        while ((fCurrValue += fIdxInc) > m_iWaveTableLength)
```
Comment (med1844) [inline src/inc/Lfo.h:57]: For each call the index should only increase by one step. Putting `+=` in while condition may cause increasing more than one step. This will lead to bugs.

**Comment by med1844** [comment]
## `.gitignore`

Please, make use of `gitignore` to keep repository clean. Look at the files changed, there are 520 files and most of them are auto generated files from IDE/CMake/etc. You can find things from CLion, .idea, xcode, cmake, and together they messed up the whole repo. They should be carefully removed.

## Result Comparison

I don't see there's any result comparison. There's only one `test_file.wav` and `matlab_graph.fig`, where's the output from the program, and where's their difference? You are also required to **test several audio files**, and currently there's only one.

## LFO parameter setter & getter

> HINT: A wavetable LFO will need its own RingBuffer. It needs to have parameters for amplitude and frequency, settable during processing.

And in `Lfo.h`, there's NO public methods to set parameters.

## Vibrato design choice

> This time, please design the class interface from scratch in src/inc/Vibrato.h. Try to design a user-friendly interface. ... Also, **at the top of the code**, provide a **comment block** with a brief **explanation about your design choices**.

And you didn't provide this.

## Tests

I doubt if vibrato tests will run as they are not attached to the gtest framework.

---

In summary there are a lot of severe issues that affect both software quality and also grading. They should be repaired ASAP.

I haven't compiled code yet, so my comments might be wrong. I will update if I need to correct / add more details and information.
